### PR TITLE
Fix consumer

### PR DESCRIPTION
**Pull request by syndbg**
* Add configurable prefetch count
* Fix RMQ channels creation
* Fix consumer cancellation not working

### CODE REVIEW

File: rabbitmq/producer.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+				Expiration: expiration,
+				Body:       body,
+			},
+		)
+		p.metric.ObserveMsgPublish(err == nil)
+
+		return stacktrace.Propagate(err, "failed to publish RMQ message")
+	}
+
+	return nil
```
Comment (slavsumup) [inline rabbitmq/producer.go:93]: Not an issue but isn't this `return nil` unreachable code?

Comment (syndbg) [reply rabbitmq/producer.go:93]: Surprisingly, no. 

We need a `default` and `return` otherwise it won't compile 😅

File: rabbitmq/consumer.go [hunk 2]
```diff
@@ -26,51 +26,85 @@ import (
...
+			_ = channel.Cancel(c.handler.GetConsumerTag(), false)
+
+			if !c.handler.WaitToConsumeInflight() {
+				_ = channel.Close()
+			}
 
-		if !c.handler.WaitToConsumeInflight() {
-			c.client.channel.Close()
+			c.logger.Info("RMQ consumer stopped.")
+			_ = c.client.Close()
```
Comment (slavsumup) [inline rabbitmq/consumer.go:94]: Would it be better if we log these potential errors, as that's still one way of handling them?

File: rabbitmq/consumer.go [hunk 3]
```diff
@@ -26,51 +26,85 @@ import (
...
 		}
-
-		<-c.done
-		c.logger.Info("handler stopped")
-		_ = c.client.Close()
 	}()
 
 	if ctx.Err() != nil {
```
Comment (slavsumup) [inline rabbitmq/consumer.go:98]: I may be missing something here but wouldn't this be nil (almost) always as the goroutine which we start above might get an error on the context a bit later whereby the execution of the parent (current) function would usually have `ctx.Err() == nil`. That is, shouldn't we get the `ctx.Err()` when we receive a value on the `<-ctx.Done()` channel?

File: rabbitmq/consumer.go [hunk 4]
```diff
@@ -80,86 +114,96 @@ func (c *RabbitMQConsumer) Run(ctx context.Context) error {
...
-			return stacktrace.Propagate(err, "handler returned error")
+func (c *Consumer) handleDeliveries(
+	ctx context.Context,
+	deliveries <-chan amqp.Delivery,
+) error {
+	for {
+		select {
+		case <-ctx.Done():
+			c.logger.Warn("RMQ handler stopping")
+			return nil
```
Comment (slavsumup) [inline rabbitmq/consumer.go:133]: Shouldn't we return `ctx.Err()` here?

File: rabbitmq/consumer.go (hunk 2 above)
Comment (syndbg) [reply rabbitmq/consumer.go:94]: Do we have if the client has been closed "incorrectly" if we're already stopping?

File: rabbitmq/consumer.go (hunk 4 above)
Comment (syndbg) [reply rabbitmq/consumer.go:133]: No, the `ctx.Err` would be pointless, since it'll be a `cancellation error`

File: rabbitmq/consumer.go (hunk 3 above)
Comment (syndbg) [reply rabbitmq/consumer.go:98]: > may be missing something here but wouldn't this be nil (almost) always as the goroutine which we start above might get an error on the context a bit later 

The context still may be canceled at this point of time. 
While the chance of this happening is very slim, the general rule-of-thumb is to check for context cancellation before committing to doing expensive I/O work, such as the communication with RMQ.

On the other hand checking before every I/O call is a bit too much, hence why we do it at this point.

File: rabbitmq/consumer.go (hunk 2 above)
Comment (slavsumup) [reply rabbitmq/consumer.go:94]: Probably not usually but maybe it's still worth it if we fail to cancel the channel, i.e. getting weird behaviour (eventually) and disregarding the errors might result in harder to find issues?

**Pull request merged**